src/linear_algebra/finsupp.lean

```diff
@@ -518,6 +518,10 @@ theorem mem_span_iff_total (s : set M) (x : M) :
   x ∈ span R s ↔ ∃ l : s →₀ R, finsupp.total s M R coe l = x :=
 (set_like.ext_iff.1 $ span_eq_range_total _ _) x
 
+lemma mem_span_range_iff_exists_finsupp {v : α → M} {x : M} :
+  x ∈ span R (range v) ↔ ∃ (c : α →₀ R), c.sum (λ i a, a • v i) = x :=
+by simp only [←finsupp.range_total, linear_map.mem_range, finsupp.total_apply]
```

Comment:
I was hoping this could simplify the proof of [`basis.mem_submodule_iff`](https://github.com/leanprover-community/mathlib/blob/cead93130da7100f8a9fe22ee210f7636a91168f/src/linear_algebra/basis.lean#L468) but it doesn't seem to help much:
```lean
lemma mem_submodule_iff {P : submodule R M} (b : basis ι R P) {x : M} :
  x ∈ P ↔ ∃ (c : ι →₀ R), x = finsupp.sum c (λ i x, x • b i) :=
begin
  conv_lhs { rw [← P.range_subtype, ← submodule.map_top, ← b.span_eq, submodule.map_span,
    ← set.range_comp] },
  simpa only [eq_comm] using finsupp.mem_span_range_iff_exists_finsupp _
end
```
If you want, you could update `mem_submodule_iff` to use this lemma (and perhaps `mem_submodule_iff'` to use `mem_span_range_iff_exists_fun`), but I think for real gains, instead we should probably extract a lemma that makes the first line easier, in another PR.

Reply:
I'll note that down for a new PR.